contracts/exchange/contracts/test/TestValidatorWallet.sol

```diff
@@ -0,0 +1,296 @@
+    ISimplifiedExchange internal _exchange;
+    /// @dev Internal state to modify.
+    uint256 internal _state = 1;
+    /// @dev What action to execute when a hash is validated .
+    mapping (bytes32=>ValidatorAction) internal _hashActions;
+    /// @dev Allowed signers for hash signature types.
+    mapping (bytes32=>address) internal _validSignerForHash;
+
+    constructor(address exchange) public {
+        _exchange = ISimplifiedExchange(exchange);
+    }
+
+    /// @dev Approves an ERC20 token to spend tokens from this address.
+    /// @param token Address of ERC20 token.
+    /// @param spender Address that will spend tokens.
+    /// @param value Amount of tokens spender is approved to spend.
+    function approveERC20(
+        address token,
+        address spender,
+        uint256 value
+    )
+        external
+    {
+        IERC20Token(token).approve(spender, value);
+    }
+
+    /// @dev Set the action to take when validating a hash.
+    /// @param hash The hash.
+    /// @param action action to take.
+    /// @param allowedSigner Signer that must be recovered with
+    ///                      `ValidateSignature` action type and `Wallet` or
+    ///                      `OrderWallet` signature types.
+    function setValidateAction(
+        bytes32 hash,
+        ValidatorAction action,
+        address allowedSigner
+    )
+        external
+    {
+        if (uint8(action) >= uint8(ValidatorAction.NTypes)) {
+            revert("UNSUPPORTED_VALIDATE_ACTION");
+        }
+        _hashActions[hash] = action;
+        _validSignerForHash[hash] = allowedSigner;
+    }
+
+    /// @dev Validates a hash with the following signature types:
+    ///      `EIP1271Wallet` and `EIP1271WalletOrder`signature types.
+    ///      The length of `data` will determine which signature type is in use.
+    /// @param data Arbitrary data. Either an Order hash or abi-encoded Order.
+    /// @param signature Signature for `data`.
+    /// @return magicValue Returns `EIP1271_MAGIC_VALUE` if the signature check succeeds.
+    function isValidSignature(
+        bytes memory data,
+        bytes memory signature
+    )
+        public
+        returns (bytes4 magicValue)
+    {
+        bytes32 hash = _getOrderHashFromEIP1271Data(data);
+        ValidatorAction action = _hashActions[hash];
+        // solhint-disable-next-line no-empty-blocks
+        if (action == ValidatorAction.Reject) {
+            // NOOP.
+        } else if (action == ValidatorAction.Accept) {
+            magicValue = EIP1271_MAGIC_VALUE;
+        } else if (action == ValidatorAction.Revert) {
+            revert(REVERT_REASON);
+        } else if (action == ValidatorAction.UpdateState) {
+            _updateState();
+        } else { // action == ValidatorAction.ValidateSignature
+            if (data.length != 32) {
+                // `data` is an abi-encoded Order.
+                LibOrder.Order memory order = _getOrderFromEIP1271Data(data);
```

Comment:
Why not use `abi.decode(data, (LibOrder.Order))` here?

Reply:
Ran into stack issues. Stuff like this might make `EIP1271OrderWallet` not a super attractive alternative to `OrderWallet`, and why I wouldn't be in favor of `EIP1271OrderWallet` supplanting `OrderWallet`. Thoughts?

Reply:
Oh, it works with `abi.decode(data, (LibOrder.Order))`. I didn't know you could do that. I was doing it the long way before (explicitly passing in all the fields).  :trollface: